walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,251 @@
+    Collection<String> mandatory = request.getAttendees();
+    long duration = request.getDuration();
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    /** 
+     * If there are conflicting mandatory attendees, then a possible range for 
+     * the meeting request starts from the end of the previous event to the 
+     * start of the conflicting event. If this time range fits the duration of 
+     * the meeting, then it is a possible time.
+     */
+    int prevEnd = 0;
+    for (Event e : eventsList){
+      Collection<String> intersection = new HashSet<>(mandatory);
+      intersection.retainAll(e.getAttendees());
+      TimeRange range = e.getWhen();
+      if (!intersection.isEmpty()) {
+        TimeRange possibleRange = TimeRange.fromStartEnd(prevEnd, range.start(), false);
+        if (duration <= possibleRange.duration()) {
+          possibleTimes.add(possibleRange);
+        }
+        if (prevEnd < range.end()) {
+          prevEnd = range.end();
+        }
+      }
+    }
+    if (prevEnd != TimeRange.END_OF_DAY + 1) {
+      possibleTimes.add(TimeRange.fromStartEnd(prevEnd, TimeRange.END_OF_DAY, true));
+    }
+    return possibleTimes;
+  }
+
+  /**
+   * Finds a specific attendee's events for the day.
+   * @param eventsList sorted list of events in ascending start time order
+   * @param attendee specific attendee to search for
+   * @return collection of events that attendee has on the schedule
+   */
+  private Collection<Event> eventFinderWithAttendee(Collection<Event> eventsList, String attendee) {
+    List<Event> eventsByAttendee = new ArrayList<Event>(); 
+
+    for (Event e : eventsList){
+      if (e.getAttendees().contains(attendee)){
+        eventsByAttendee.add(e);
+      }
+    }
+    return eventsByAttendee;
+  }
+
+  /**
+   * Split large range into smaller pieces excluding overlapping events. If an 
+   * event overlaps with the range, then the range can be broken into the subset
+   * of the range that occurs before the first event and additional gaps between
+   * events. We take the ranges between the end of the previous event and the 
+   * start of the next event, adding a range from the last event's end time to 
+   * the end of the range if necessary.
+   * Range:  |-------------|
+   * Event:       |---|
+   * Return: |----|   |----|
+   * @param eventsList sorted list of events in ascending start time order
+   * @param range large range to split into smaller ranges
+   * @param duration how long the request is for
+   * @return collection of TimeRanges that was part of the large range without event overlaps
+   */
+  private Collection<TimeRange> rangeSplitByNestedEvents(Collection<Event> eventsList, TimeRange range, long duration) {
+    List<TimeRange> splitEvents = new ArrayList<TimeRange>(); 
+    int prevEndTime = range.start();
+
+    for (Event e : eventsList) {
+      if (range.overlaps(e.getWhen())) {
+        if (range.start() <= e.getWhen().start()) {
```

Comment:
This if statement is a bit confusing to me.
The eventList is sorted by start time in ascending order, which means you don't have to compare range.Start() with event.Start() for each event.  Looks more suitable to move the if condition outside of the for loop.